File: backend/ai_logic.py
"""
AMAN AI Logic — Sequencing decisions ONLY.
No movement. No state transitions. Those are in physics.py.

EUROCONTROL AMAN Rules:
- Aircraft sorted by ETT (distance/speed)
- Sequence numbers are always stable (committed aircraft first)
- Only ONE aircraft LANDING/ROLLOUT per runway at a time
- If runway free → next sequence aircraft must directly land (when ≤ 6 NM)
- Speed reduction FIRST, holding is LAST RESORT
- Routes assigned ONCE — never reset every tick
"""

import math
from typing import List, Tuple, Optional, Dict
from models import (
    Aircraft, AircraftStatus, WakeCategory, WAKE_SEPARATION_SEC,
    haversine_distance, calculate_bearing, offset_position,
)
from state import atc_state
from config import (
    RUNWAYS, OPTIMIZATION_ZONE_NM, SEPARATION_TIME_SEC,
    SPEED_PROFILES, FINAL_APPROACH_START_NM, MIN_SEPARATION_NM,
    DELAY_SMALL_THRESHOLD_SEC, DELAY_MEDIUM_THRESHOLD_SEC, DELAY_HOLD_RELEASE_SEC,
)


def run_sequencing() -> None:
    """Main entry — called every tick BEFORE physics."""
    aircraft_list = atc_state.get_active_aircraft()
    if not aircraft_list:
        return

    # Refresh distances from actual positions
    for ac in aircraft_list:
        ac.distance_to_threshold = ac.remaining_distance_nm()

    # Group by runway (only APPROACH/HOLDING/LANDING/ROLLOUT)
    groups: Dict[str, List[Aircraft]] = {rwy: [] for rwy in RUNWAYS}
    for ac in aircraft_list:
        if ac.runway in groups and ac.status != AircraftStatus.LANDED:
            groups[ac.runway].append(ac)

    for runway_id, acs in groups.items():
        _sequence_runway(runway_id, acs)

    _detect_conflicts([ac for ac in aircraft_list if ac.status != AircraftStatus.LANDED])

    # Write back to state
    for ac in aircraft_list:
        atc_state.aircraft[ac.id] = ac


def _get_separation(leader: Aircraft, follower: Aircraft) -> float:
    """Wake turbulence separation in seconds."""
    return WAKE_SEPARATION_SEC.get(
        leader.wake_category, {}
    ).get(follower.wake_category, SEPARATION_TIME_SEC)


def _sequence_runway(runway_id: str, aircraft: List[Aircraft]) -> None:
    """
    EUROCONTROL AMAN sequencing for a single runway.
    Committed (LANDING/ROLLOUT) → first. Sequenceable (APPROACH/HOLDING) → sorted by ETT.
    """
    if not aircraft:
        return
    rc = RUNWAYS.get(runway_id)
    if not rc:
        return

    threshold = (rc["threshold_lat"], rc["threshold_lon"])

    # Split committed vs. sequenceable
    committed = [ac for ac in aircraft
                 if ac.status in (AircraftStatus.LANDING, AircraftStatus.ROLLOUT)]
    sequenceable = [ac for ac in aircraft
                    if ac.status in (AircraftStatus.APPROACH, AircraftStatus.HOLDING)]

    # Sort sequenceable by ETT (ascending = closest first)
    for ac in sequenceable:
        ac._ett = ac.compute_ett()
    sequenceable.sort(key=lambda a: a._ett)

    # Assign stable sequence numbers: committed first, then ETT order
    for idx, ac in enumerate(committed):
        ac.sequence_position = idx + 1
        ac.queue_position = 0

    # Determine if runway is occupied by a committed aircraft
    runway_occupied = len(committed) > 0

    # Estimate when runway will next be free
    base_rta = atc_state.simulation_time
    if runway_occupied:
        rollout = [ac for ac in committed if ac.status == AircraftStatus.ROLLOUT]
        landing = [ac for ac in committed if ac.status == AircraftStatus.LANDING]
        if rollout:
            # Rollout aircraft: ~ 30 seconds to vacate
            base_rta = atc_state.simulation_time + 30
        elif landing:
            ac_l = landing[0]
            time_to_thresh = (ac_l.distance_to_threshold / max(ac_l.speed, 1)) * 3600
            base_rta = atc_state.simulation_time + time_to_thresh + 35  # + rollout
    else:
        # Check separation from last cleared aircraft
        rwy = atc_state.get_runway(runway_id)
        if rwy and rwy.last_cleared_time > 0:
            since = atc_state.simulation_time - rwy.last_cleared_time
            if since < SEPARATION_TIME_SEC:
                base_rta = atc_state.simulation_time + (SEPARATION_TIME_SEC - since)

    # Assign RTA slots to sequenceable aircraft
    prev_ac: Optional[Aircraft] = committed[-1] if committed else None
    prev_rta = base_rta

    for idx, ac in enumerate(sequenceable):
        seq_num = len(committed) + idx + 1
        ac.sequence_position = seq_num
        ac.queue_position = idx

        ett = ac._ett
        # Absolute RTA must be at least (simulation_time + ETT) and satisfy separation
        if idx == 0:
            rta_abs = max(base_rta, atc_state.simulation_time + ett)
        else:
            sep = _get_separation(prev_ac, ac) if prev_ac else SEPARATION_TIME_SEC
            rta_abs = max(prev_rta + sep, atc_state.simulation_time + ett)

        rta_from_now = max(0.0, rta_abs - atc_state.simulation_time)
        ac.rta = rta_from_now
        delay = max(0.0, rta_from_now - ett)
        ac.predicted_delay = delay
        atc_state.record_delay(ac.id, delay)

        # Compute optimal speed to meet RTA
        ac.optimal_speed = _compute_optimal_speed(ac, rta_from_now)

        prev_rta = rta_abs
        prev_ac = ac

        # ── LANDING CLEARANCE DECISION ──────────────────────────────────────
        # Skip if already cleared for landing
        if getattr(ac, '_landing_cleared', False):
            continue

        # EUROCONTROL rule: runway free + first in sequence + ≤ FINAL_APPROACH_START_NM
        if (idx == 0
                and not runway_occupied
                and ac.distance_to_threshold <= FINAL_APPROACH_START_NM):
            _clear_landing(ac, runway_id, threshold, rc)
            runway_occupied = True

        elif ac.distance_to_threshold > OPTIMIZATION_ZONE_NM:
            # Far out — normal approach, speed reduction only
            _set_approach(ac, runway_id, seq_num)

        else:
            # Inside TMA but cannot land yet — absorb delay
            _absorb_delay(ac, runway_id, threshold, delay, seq_num)


def _compute_optimal_speed(ac: Aircraft, rta_from_now: float) -> float:
    """Calculate speed needed to meet RTA without holding."""
    if rta_from_now <= 0 or ac.distance_to_threshold <= 0:
        return ac.speed
    required_kts = (ac.distance_to_threshold / rta_from_now) * 3600
    profile = SPEED_PROFILES["APPROACH"]
    return max(profile["min"], min(profile["max"], required_kts))


# =============================================================================
# STATE ASSIGNMENTS
# =============================================================================

def _set_approach(ac: Aircraft, runway_id: str, seq_num: int = 0) -> None:
    """Normal approach — speed control only, no route override."""
    if ac.status in (AircraftStatus.LANDING, AircraftStatus.ROLLOUT):
        return
    # If holding with active route, let it finish before switching back
    if (ac.status == AircraftStatus.HOLDING
            and ac.route
            and ac.current_waypoint_index < len(ac.route)):
        return
    ac.status = AircraftStatus.APPROACH
    ac.aman_heading_override = False
    ac.delay_mode = "speed"

    # ATC instruction: speed reduction
    spd = int(ac.optimal_speed) if ac.optimal_speed > 0 else int(ac.speed)
    if seq_num > 0:
        ac.instruction = f"{ac.callsign}, reduce speed {spd} knots, sequence #{seq_num}."
    else:
        ac.instruction = f"{ac.callsign}, reduce speed {spd} knots."


def _clear_landing(ac: Aircraft, runway_id: str, threshold: Tuple, rc: dict) -> None:
    """
    Clear for landing — EUROCONTROL AMAN landing clearance.
    Assigns route ONCE: current position → threshold → exit.
    Snaps to threshold on touchdown (in physics.py).
    """
    ac.status = AircraftStatus.LANDING
    ac.predicted_delay = 0
    ac.delay_mode = "speed"
    ac.aman_heading_override = True
    ac._landing_cleared = True
    ac.instruction = f"{ac.callsign}, runway {runway_id} cleared to land. Wind calm."

    # Generate ILS approach route: intermediate points → threshold
    route = []
    steps = max(4, int(ac.distance_to_threshold / 1.0))
    for i in range(1, steps + 1):
        t = i / steps
        lat = ac.lat + (threshold[0] - ac.lat) * t
        lon = ac.lon + (threshold[1] - ac.lon) * t
        alt = max(0.0, ac.altitude * (1.0 - t))
        route.append([lat, lon, alt])
    # Final point: exactly at threshold, altitude 0
    route[-1] = [threshold[0], threshold[1], 0.0]

    # Rollout segment: threshold → runway exit (4 intermediate points)
    exit_lat, exit_lon = rc["exit_lat"], rc["exit_lon"]
    for i in range(1, 5):
        t = i / 4
        lat = threshold[0] + (exit_lat - threshold[0]) * t
        lon = threshold[1] + (exit_lon - threshold[1]) * t
        route.append([lat, lon, 0.0])

    ac.route = route
    ac.current_waypoint_index = 0

    # Lock the runway
    atc_state.set_runway_occupied(runway_id, ac)


def _absorb_delay(
    ac: Aircraft,
    runway_id: str,
    threshold: Tuple,
    delay: float,
    seq_num: int,
) -> None:
    """
    3-tier delay absorption strategy.
    Tier 1: Speed reduction (primary — preferred)
    Tier 2: Trombone vectoring (medium delay, 1–3 min)
    Tier 3: Holding pattern (last resort, > 3 min)
    """

    # ── Release from holding if delay dissipated ────────────────────────────
    if ac.delay_mode in ("holding", "trombone") and delay < DELAY_HOLD_RELEASE_SEC:
        ac.delay_mode = "speed"
        ac.status = AircraftStatus.APPROACH
        ac.aman_heading_override = False
        ac.route = []
        ac.current_waypoint_index = 0
        spd = int(ac.optimal_speed) if ac.optimal_speed > 0 else int(ac.speed)
        ac.instruction = (
            f"{ac.callsign}, #{seq_num}. Holding cancelled, resume approach. "
            f"Reduce speed {spd} knots."
        )
        atc_state.record_holding_exit(ac.id)
        return

    # ── Tier 1: Speed reduction ──────────────────────────────────────────────
    profile = SPEED_PROFILES["APPROACH"]
    min_speed_nm_per_sec = profile["min"] / 3600.0
    time_at_min_speed = (
        (ac.distance_to_threshold / profile["min"]) * 3600
        if profile["min"] > 0 else 0
    )
    max_speed_absorption_sec = time_at_min_speed - ac._ett  # extra time by flying min speed

    if delay <= max(max_speed_absorption_sec, DELAY_SMALL_THRESHOLD_SEC):
        ac.delay_mode = "speed"
        if ac.status != AircraftStatus.HOLDING:
            ac.status = AircraftStatus.APPROACH
        ac.aman_heading_override = False
        spd = int(ac.optimal_speed) if ac.optimal_speed > 0 else int(ac.speed)
        ac.instruction = (
            f"{ac.callsign}, #{seq_num}. Reduce speed {spd} knots. "
            f"{int(delay)}s delay."
        )
        return

    # ── Tier 2: Trombone vectoring ───────────────────────────────────────────
    if delay <= DELAY_MEDIUM_THRESHOLD_SEC:
        ac.delay_mode = "trombone"
        # Only generate route if none active or completed
        if not ac.route or ac.current_waypoint_index >= len(ac.route) - 1:
            ac.route = _generate_trombone(ac, threshold, delay, runway_id)
            ac.current_waypoint_index = 0
        ac.status = AircraftStatus.HOLDING
        ac.aman_heading_override = True
        ac.instruction = (
            f"{ac.callsign}, #{seq_num}. Vectoring for spacing. "
            f"{round(delay / 60, 1)} min delay."
        )
        return

    # ── Tier 3: Holding (last resort) ────────────────────────────────────────
    prev_mode = ac.delay_mode
    ac.delay_mode = "holding"
    if (not ac.route
            or ac.current_waypoint_index >= len(ac.route) - 1
            or prev_mode != "holding"):
        ac.route = _generate_trombone(ac, threshold, delay, runway_id)
        ac.current_waypoint_index = 0
        atc_state.record_holding_entry(ac.id)

    ac.status = AircraftStatus.HOLDING
    ac.aman_heading_override = True
    ac.instruction = (
        f"{ac.callsign}, #{seq_num}. Hold present position. "
        f"{round(delay / 60, 1)} min delay. "
        f"Maintain {int(ac.altitude)} ft."
    )


def _generate_trombone(
    ac: Aircraft,
    threshold: Tuple,
    delay_sec: float,
    runway_id: str,
) -> List[List[float]]:
    """
    Smooth trombone/racetrack route.
    lateral offset → downwind → base → FAF → threshold
    Does NOT loop — ends at runway threshold.
    """
    rc = RUNWAYS.get(runway_id)
    if not rc:
        return [[threshold[0], threshold[1], 50.0]]

    rwy_hdg = rc["heading"]
    hold_speed = SPEED_PROFILES["HOLDING"]["target"]
    extra_nm = (delay_sec / 3600.0) * hold_speed
    downwind_nm = max(3.0, min(25.0, extra_nm / 2.0))

    offset_dir = 1 if runway_id == "09L" else -1
    lat_bearing = (rwy_hdg + 90 * offset_dir) % 360
    opp_lat_bearing = (lat_bearing + 180) % 360
    approach_bearing = (rwy_hdg + 180) % 360
    lat_offset_nm = 3.0

    faf = offset_position(threshold[0], threshold[1], approach_bearing, FINAL_APPROACH_START_NM)
    hold_alt = min(ac.altitude, 5000.0)

    route = []
    # 1: lateral abeam offset
    abeam = offset_position(ac.lat, ac.lon, lat_bearing, lat_offset_nm)
    route.append([abeam[0], abeam[1], hold_alt])
    # 2: downwind leg
    dw = offset_position(abeam[0], abeam[1], approach_bearing, downwind_nm)
    route.append([dw[0], dw[1], 4000.0])
    # 3: base turn
    base = offset_position(dw[0], dw[1], opp_lat_bearing, lat_offset_nm)
    route.append([base[0], base[1], 3500.0])
    # 4: Final Approach Fix
    route.append([faf[0], faf[1], 2500.0])
    # 5: Threshold
    route.append([threshold[0], threshold[1], 50.0])

    return route


# =============================================================================
# CONFLICT DETECTION
# =============================================================================

def _detect_conflicts(aircraft: List[Aircraft]) -> None:
    """Minimum separation check — sets has_conflict flag on model."""
    for i, ac1 in enumerate(aircraft):
        min_sep = float("inf")
        conflict = False
        for j, ac2 in enumerate(aircraft):
            if i == j:
                continue
            h = haversine_distance(ac1.lat, ac1.lon, ac2.lat, ac2.lon)
            v = abs(ac1.altitude - ac2.altitude)
            if h < MIN_SEPARATION_NM and v < 1000:
                conflict = True
            if h < min_sep:
                min_sep = h
        ac1.has_conflict = conflict
        ac1.safety_percent = (
            min(100.0, max(0.0, (min_sep / MIN_SEPARATION_NM) * 100.0))
            if len(aircraft) > 1 else 100.0
        )


# =============================================================================
# STATE TRANSITIONS (post-physics cleanup)
# =============================================================================

def process_state_transitions() -> None:
    """
    Post-physics cleanup.
    Remove LANDED aircraft from state.
    NOTE: Runway is already cleared by physics.py when aircraft exits (ROLLOUT→LANDED).
    Do NOT call clear_runway() here — that would double-count the landing event.
    """
    # Iterate over ALL aircraft in state to find those that are LANDED
    to_remove = [
        ac_id
        for ac_id, ac in atc_state.aircraft.items()
        if ac.status == AircraftStatus.LANDED
    ]
    for ac_id in to_remove:
        atc_state.remove_aircraft(ac_id)

File: supabase/functions/_shared/ai_logic.ts
/**
 * AMAN AI Logic - The Brain (Async version)
 * 
 * This is the CORE AMAN decision engine.
 * 
 * KEY RULES:
 * 1. 20 NM Optimization Zone - Only aircraft inside 20 NM are sequenced
 * 2. Distance-based priority - NOT FIFO, closest aircraft lands first
 * 3. One landing at a time per runway
 * 4. Holding patterns for all other aircraft inside 20 NM
 */

import { 
  Aircraft, 
  RunwayId, 
  Waypoint,
  haversineDistance,
  remainingDistanceNM,
  updateETA
} from './models.ts';
import {
  getAircraftList,
  getRunwayState,
  updateAircraft,
  setRunwayOccupied,
  setRunwayClearanceLock,
  AIRPORT
} from './state.ts';

// AMAN Constants
const OPTIMIZATION_ZONE_NM = 20;   // Aircraft inside this zone are sequenced
const MIN_SEPARATION_NM = 3;       // Minimum horizontal separation
const MIN_SEPARATION_TIME = 90;    // Seconds between landings on same runway

// Speed profiles (knots) - used by physics engine
export const SPEED_PROFILES = {
  APPROACHING: { min: 220, max: 250 },  // Outside 20 NM
  HOLDING: { min: 170, max: 190 },      // Inside 20 NM, waiting
  LANDING: { min: 130, max: 150 },      // Final approach
  ROLLOUT: { min: 0, max: 60 },         // On runway
};

/**
 * Get appropriate speed for aircraft status
 */
export function getTargetSpeed(status: string, distanceToThreshold: number): number {
  switch (status) {
    case 'LANDING':
      return SPEED_PROFILES.LANDING.max; // 150 knots on final
    case 'ROLLOUT':
      return SPEED_PROFILES.ROLLOUT.max;
    case 'HOLDING':
      return SPEED_PROFILES.HOLDING.max; // 190 knots in holding
    case 'APPROACHING':
    default:
      // Slow down as we get closer
      if (distanceToThreshold < 25) {
        return SPEED_PROFILES.APPROACHING.min; // 220 knots
      }
      return SPEED_PROFILES.APPROACHING.max; // 250 knots
  }
}

/**
 * Main AMAN sequencing function - called every tick
 * Implements proper 20 NM optimization zone with distance-based priority
 */
export async function runSequencing(): Promise<void> {
  const aircraft = await getAircraftList();
  if (aircraft.length === 0) return;
  
  // Group by runway
  const byRunway = new Map<RunwayId, Aircraft[]>();
  byRunway.set('09L', []);
  byRunway.set('09R', []);
  
  for (const ac of aircraft) {
    if (ac.status !== 'LANDED') {
      byRunway.get(ac.runway)?.push(ac);
    }
  }
  
  // Process each runway independently
  for (const [runwayId, runwayAircraft] of byRunway) {
    await processRunwaySequence(runwayId, runwayAircraft);
  }
  
  // Check for conflicts between all aircraft
  await detectConflicts(aircraft);
}

/**
 * Process sequencing for a single runway
 * CRITICAL: Implements 20 NM optimization zone with distance-based priority
 */
async function processRunwaySequence(runwayId: RunwayId, aircraft: Aircraft[]): Promise<void> {
  if (aircraft.length === 0) return;
  
  const runwayState = await getRunwayState(runwayId);
  if (!runwayState) return;
  
  // Calculate distance for each aircraft and update ETA
  const aircraftWithDistance: Array<{ aircraft: Aircraft; distance: number }> = [];
  
  for (const ac of aircraft) {
    const distance = remainingDistanceNM(ac);
    const targetSpeed = getTargetSpeed(ac.status, distance);
    const eta = (distance / targetSpeed) * 3600; // seconds
    
    await updateAircraft(ac.id, { 
      distanceToThreshold: distance,
      eta,
      speed: targetSpeed, // Update speed based on status
    });
    
    aircraftWithDistance.push({ aircraft: ac, distance });
  }
  
  // Sort ALL aircraft by remaining distance (closest first) - NOT FIFO!
  aircraftWithDistance.sort((a, b) => a.distance - b.distance);
  
  // Assign sequence numbers based on distance (for display purposes)
  for (let i = 0; i < aircraftWithDistance.length; i++) {
    await updateAircraft(aircraftWithDistance[i].aircraft.id, { 
      sequenceNumber: i + 1 
    });
  }
  
  // Separate aircraft by zone
  const insideZone = aircraftWithDistance.filter(a => a.distance <= OPTIMIZATION_ZONE_NM);
  const outsideZone = aircraftWithDistance.filter(a => a.distance > OPTIMIZATION_ZONE_NM);
  
  // Check if runway is occupied (LANDING or ROLLOUT in progress)
  const hasActiveLanding = aircraft.some(ac => 
    ac.status === 'LANDING' || ac.status === 'ROLLOUT'
  );
  
  // Process aircraft OUTSIDE 20 NM zone - normal approach, no holding
  for (const { aircraft: ac, distance } of outsideZone) {
    if (ac.status === 'HOLDING') {
      // Aircraft moved outside zone - return to approach
      await updateAircraft(ac.id, {
        status: 'APPROACHING',
        route: [],
        currentWaypointIndex: 0,
        atcInstruction: `${ac.callsign}, continue approach. Descend to FL100.`,
      });
    } else if (ac.status === 'APPROACHING') {
      // Normal approach - no action needed
      await updateAircraft(ac.id, {
        atcInstruction: `${ac.callsign}, radar contact. Continue approach runway ${runwayId}.`,
      });
    }
  }
  
  // Process aircraft INSIDE 20 NM zone - optimization logic
  for (let i = 0; i < insideZone.length; i++) {
    const { aircraft: ac, distance } = insideZone[i];
    const isNearest = i === 0;
    
    // Skip aircraft already committed to landing
    if (ac.status === 'LANDING' || ac.status === 'ROLLOUT' || ac.status === 'LANDED') {
      continue;
    }
    
    if (isNearest && !hasActiveLanding && !runwayState.clearance_lock) {
      // NEAREST aircraft inside 20 NM - clear for landing!
      await clearForLanding(ac, runwayId, distance);
    } else {
      // ALL OTHER aircraft inside 20 NM - must hold
      if (ac.status !== 'HOLDING') {
        await assignHolding(ac, i, insideZone.length, distance);
      } else {
        // Update holding instruction with current position in queue
        await updateAircraft(ac.id, {
          atcInstruction: `${ac.callsign}, #${i + 1} in sequence. Hold present position. Expect ${Math.ceil((i * MIN_SEPARATION_TIME) / 60)} min delay.`,
        });
      }
    }
  }
}

/**
 * Clear aircraft for landing
 */
async function clearForLanding(aircraft: Aircraft, runwayId: RunwayId, distance: number): Promise<void> {
  // Lock runway to prevent double clearance
  await setRunwayClearanceLock(runwayId, true);
  
  // Generate direct approach route to threshold
  const threshold = AIRPORT.runwayThresholds[runwayId];
  const finalRoute = generateFinalApproachRoute(
    aircraft.lat, aircraft.lon, aircraft.altitude,
    threshold.lat, threshold.lon
  );
  
  await updateAircraft(aircraft.id, {
    status: 'LANDING',
    route: finalRoute,
    currentWaypointIndex: 0,
    speed: SPEED_PROFILES.LANDING.max, // 150 knots
    atcInstruction: `${aircraft.callsign}, runway ${runwayId} cleared to land. Wind calm.`,
    hasConflict: false,
  });
  
  // Mark runway as occupied
  await setRunwayOccupied(runwayId, true, aircraft.id);
}

/**
 * Assign holding with delay absorption route
 * 
 * Uses elongated paths that absorb delay through track miles
 * NO circular holding patterns!
 */
async function assignHolding(
  aircraft: Aircraft, 
  sequencePosition: number, 
  totalInSequence: number,
  currentDistance: number
): Promise<void> {
  // Calculate required delay based on sequence position
  const baseDelay = MIN_SEPARATION_TIME; // seconds between each aircraft
  const requiredDelay = sequencePosition * baseDelay;
  
  // Generate delay absorption route
  const holdingRoute = generateDelayAbsorptionRoute(
    aircraft,
    requiredDelay,
    sequencePosition,
    currentDistance
  );
  
  const estimatedHoldTime = Math.ceil(requiredDelay / 60);
  
  await updateAircraft(aircraft.id, {
    status: 'HOLDING',
    route: holdingRoute,
    currentWaypointIndex: 0,
    speed: SPEED_PROFILES.HOLDING.max, // 190 knots
    atcInstruction: `${aircraft.callsign}, #${sequencePosition + 1} in sequence. Expect ${estimatedHoldTime} minute delay. Fly heading ${Math.round(aircraft.heading)}°, vectors for spacing.`,
  });
}

/**
 * Generate delay absorption route
 * 
 * Creates an elongated path (NOT circular) that:
 * 1. Starts from aircraft's current position
 * 2. Extends laterally to absorb required delay as track miles
 * 3. Gradually converges toward the runway
 */
function generateDelayAbsorptionRoute(
  aircraft: Aircraft,
  delaySeconds: number,
  sequencePosition: number,
  currentDistance: number
): Waypoint[] {
  const route: Waypoint[] = [];
  const threshold = AIRPORT.runwayThresholds[aircraft.runway];
  
  // Calculate track miles needed to absorb delay at holding speed
  const holdingSpeed = SPEED_PROFILES.HOLDING.max; // 190 knots
  const trackMilesNeeded = (delaySeconds / 3600) * holdingSpeed;
  
  // Current position
  const startLat = aircraft.lat;
  const startLon = aircraft.lon;
  const startAlt = aircraft.altitude;
  
  // Offset based on sequence position and runway
  const baseOffset = 0.03;
  const positionOffset = sequencePosition * 0.015;
  const lateralOffset = baseOffset + positionOffset;
  
  // Alternate north/south based on runway to separate traffic
  const isNorthOffset = aircraft.runway === '09L';
  const offsetDir = isNorthOffset ? 1 : -1;
  
  // Waypoint 1: Initial vector away from approach path
  const wp1Lat = startLat + (offsetDir * lateralOffset * 0.5);
  const wp1Lon = startLon - 0.03;
  route.push({ lat: wp1Lat, lon: wp1Lon, alt: startAlt });
  
  // Waypoint 2: Downwind leg (parallel to runway, offset)
  const extendedLength = Math.min(0.1, trackMilesNeeded / 150);
  const wp2Lat = startLat + (offsetDir * lateralOffset);
  const wp2Lon = startLon - 0.06 - extendedLength;
  route.push({ lat: wp2Lat, lon: wp2Lon, alt: Math.max(startAlt - 500, 3500) });
  
  // Waypoint 3: Base turn
  const wp3Lat = threshold.lat + (offsetDir * lateralOffset * 0.6);
  const wp3Lon = threshold.lon - 0.06;
  route.push({ lat: wp3Lat, lon: wp3Lon, alt: 3000 });
  
  // Waypoint 4: Final approach intercept
  const wp4Lat = threshold.lat + (offsetDir * 0.015);
  const wp4Lon = threshold.lon - 0.035;
  route.push({ lat: wp4Lat, lon: wp4Lon, alt: 2500 });
  
  // Waypoint 5: Final approach fix (stable approach point)
  const fafLat = threshold.lat;
  const fafLon = threshold.lon - 0.025;
  route.push({ lat: fafLat, lon: fafLon, alt: 2000 });
  
  return route;
}

/**
 * Generate final approach route (direct to runway)
 */
function generateFinalApproachRoute(
  startLat: number, startLon: number, startAlt: number,
  thresholdLat: number, thresholdLon: number
): Waypoint[] {
  const route: Waypoint[] = [];
  const numPoints = 15;
  
  // Runway end (touchdown zone + rollout)
  const runwayEndLon = AIRPORT.runwayEndLon;
  
  // Generate smooth descent to threshold
  for (let i = 0; i <= numPoints; i++) {
    const t = i / numPoints;
    
    // First 70%: approach to threshold
    if (t <= 0.7) {
      const approachT = t / 0.7;
      route.push({
        lat: startLat + (thresholdLat - startLat) * approachT,
        lon: startLon + (thresholdLon - startLon) * approachT,
        alt: Math.max(startAlt * (1 - approachT * 0.95), 50),
      });
    } else {
      // Last 30%: rollout on runway
      const rolloutT = (t - 0.7) / 0.3;
      route.push({
        lat: thresholdLat,
        lon: thresholdLon + (runwayEndLon - thresholdLon) * rolloutT,
        alt: 0,
      });
    }
  }
  
  return route;
}

/**
 * Detect conflicts between aircraft
 */
async function detectConflicts(aircraft: Aircraft[]): Promise<void> {
  for (let i = 0; i < aircraft.length; i++) {
    const ac1 = aircraft[i];
    let hasConflict = false;
    let minSeparation = Infinity;
    
    for (let j = 0; j < aircraft.length; j++) {
      if (i === j) continue;
      const ac2 = aircraft[j];
      
      // Calculate separation
      const horizontalDist = haversineDistance(ac1.lat, ac1.lon, ac2.lat, ac2.lon);
      const verticalDist = Math.abs(ac1.altitude - ac2.altitude);
      
      // Check for conflict (within 3NM horizontal AND 1000ft vertical)
      if (horizontalDist < MIN_SEPARATION_NM && verticalDist < 1000) {
        hasConflict = true;
      }
      
      // Track minimum separation for safety score
      const separation = Math.sqrt(
        horizontalDist * horizontalDist + 
        (verticalDist / 1000) * (verticalDist / 1000)
      );
      minSeparation = Math.min(minSeparation, separation);
    }
    
    // Calculate safety percentage (0-100)
    const safetyPercent = aircraft.length > 1 
      ? Math.min(100, Math.round((minSeparation / 5) * 100))
      : 100;
    
    if (hasConflict) {
      await updateAircraft(ac1.id, { 
        hasConflict: true, 
        safetyPercent,
        atcInstruction: `${ac1.callsign}, TRAFFIC ALERT! Maintain present altitude and heading.`
      });
    } else {
      await updateAircraft(ac1.id, { 
        hasConflict: false, 
        safetyPercent
      });
    }
  }
}

/**
 * Handle runway vacation (called when aircraft completes rollout)
 */
export async function vacateRunway(aircraftId: string, runwayId: RunwayId): Promise<void> {
  await setRunwayOccupied(runwayId, false, null);
  await setRunwayClearanceLock(runwayId, false);
  
  // The next aircraft in sequence will be cleared on next tick
  // This triggers re-sequencing based on distance
}
